README.md
# Django-invitations - Generic invitations app

[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

## About

> This a fork of [django-invitations](https://github.com/bee-keeper/django-invitations). By March 2022, the original project wasn't being updated and I needed it to be compactible with Django 4.0+ for my own projects.

Generic invitations solution with adaptable backend and support for django-allauth.  All emails and messages are fully customisable.

Originally written as an invitations solution for the excellent [django-allauth](https://github.com/pennersr/django-allauth), this app has been refactored to remove the allauth dependency whilst retaining 100% backwards compatibility.

## Installation

```bash
pip install django-invitations

# Add to settings.py, INSTALLED_APPS
'invitations',

# Append to urls.py
path('invitations/', include('invitations.urls', namespace='invitations')),

# Run migrations

python manage.py migrate
```

## Usage

There are two primary ways to use `django-invitations` described below.

Generic Invitation flow:

* Priviledged user invites prospective user by email (via either Django admin, form post, JSON post or programmatically)
* User receives invitation email with confirmation link
* User clicks link and is redirected to a preconfigured url (default is accounts/signup)

Allauth Invitation flow:

* As above but..
* User clicks link, their email is confirmed and they are redirected to signup
* The signup URL has the email prefilled and upon signing up the user is logged into the site

Further details can be found in the following sections.

### Allauth Integration

As above but note that invitations must come after allauth in the INSTALLED_APPS

```python
# Add to settings.py
ACCOUNT_ADAPTER = 'invitations.models.InvitationsAdapter'
```

### Sending Invites

First import the model:

```python
from invitations.utils import get_invitation_model
```

Make an instance of the model:

```python
Invitation = get_invitation_model()
```

Then finally pass the recipient to the model and send.

```python
# inviter argument is optional
invite = Invitation.create('email@example.com', inviter=request.user)
invite.send_invitation(request)
```

To send invites via django admin, just add an invite and save.

### Bulk Invites

Bulk invites are supported via JSON.  Post a list of comma separated emails to the dedicated URL and Invitations will return a data object containing a list of valid and invalid invitations.

### Testing

`python manage.py test` or `tox`

### Additional Configuration

* `INVITATIONS_INVITATION_EXPIRY` (default=`3`)

    Integer.  How many days before the invitation expires.

* `INVITATIONS_INVITATION_ONLY` (default=`False`)

    Boolean.  If the site is invite only, or open to all (only relevant when using allauth).

* `INVITATIONS_CONFIRM_INVITE_ON_GET` (default=`True`)

    Boolean. If confirmations can be accepted via a `GET` request.

* `INVITATIONS_ACCEPT_INVITE_AFTER_SIGNUP` (default=`False`)

    Boolean. If `True`, invitations will be accepted after users finish signup.
    If `False`, invitations will be accepted right after the invitation link is clicked.
    Note that this only works with Allauth for now, which means `ACCOUNT_ADAPTER` has to be
    `'invitations.models.InvitationsAdapter'`.

* `INVITATIONS_GONE_ON_ACCEPT_ERROR` (default=`True`)

    Boolean. If `True`, return an HTTP 410 GONE response if the invitation key
    is invalid, or the invitation is expired or previously accepted when
    accepting an invite. If `False`, display an error message and redirect on
    errors:

  * Redirects to `INVITATIONS_SIGNUP_REDIRECT` on an expired key
  * Otherwise, redirects to `INVITATIONS_LOGIN_REDIRECT` on other errors.

* `INVITATIONS_ALLOW_JSON_INVITES` (default=`False`)

    Expose a URL for authenticated posting of invitees

* `INVITATIONS_SIGNUP_REDIRECT` (default=`'account_signup'`)

    URL name of your signup URL.

* `INVITATIONS_LOGIN_REDIRECT` (default=`LOGIN_URL` from Django settings)

    URL name of your login URL.

* `INVITATIONS_ADAPTER` (default=`'invitations.adapters.BaseInvitationsAdapter'`)

    Used for custom integrations. Set this to `ACCOUNT_ADAPTER` if using django-allauth.

* `INVITATIONS_EMAIL_MAX_LENGTH` (default=`254`)

    If set to `None` (the default), invitation email max length will be set up to 254. Set this to an integer value to set up a custome email max length value.

* `INVITATIONS_EMAIL_SUBJECT_PREFIX` (default=`None`)

    If set to `None` (the default), invitation email subjects will be prefixed with the name of the current Site in brackets (such as `[example.com]`). Set this to a string to for a custom email subject prefix, or an empty string for no prefix.

* `INVITATIONS_INVITATION_MODEL` (default=`invitations.Invitation`)

    App registry path of the invitation model used in the current project, for customization purposes.

### Signals

The following signals are emitted:

* `invite_url_sent`
* `invite_accepted`

### Management Commands

Expired and accepted invites can be cleared as so:

`python manage.py clear_expired_invitations`


CHANGELOG.md
# Changelog

## 1.10 (2022-03-28)

(not yet released)

- markdownlint style formatting [MrCordeiro]
- black style formatting [MrCordeiro]
- Remove support for python 3.5, 3.6. Add support for python 3.9, 3.10 [MrCordeiro]
- Remove support for Django 1.1, 2.2. Add support for Django 4.0 [MrCordeiro]

## 1.9 (2017-02-11)

- Added get_signup_redirect to allow custom implementations by subclasses

- Fixed invitation form displaying "None" when first displayed

- Fixed deprecation warnings

- Added get_signup_redirect to allow custom implementations by subclasses

- Fixed flake8 errors

- Import reverse from django.urls if available, otherwise fall back to old import

- Set ForeignKey field to explicitly cascade on deletion

- flake8 styling formatting

- Add email max length setting

## 1.8 (2016-10-19)

- [TO BE ADDED]

## 1.7 (2016-02-10)

- 1.7. [bee-keeper]

- Merge pull request #34 from percipient/dont-404. [bee-keeper]

  Display a message on expired/erroneous/accepted keys.

- Fix flake8 issues. [Patrick Cloke]

- Fix a formatting mistake in the README. [Patrick Cloke]

- Display a message on expired/erroneous/accepted keys. Related to #25.
  [Patrick Cloke]

- Merge pull request #33 from percipient/readme-settings. [bee-keeper]

  Fix-up the settings in the README

- Update the INVITATIONS_ADAPTER setting. [Patrick Cloke]

- Update the readme with all the settings. [Patrick Cloke]

- Removed uneeded check on send invite. [bee_keeper]

- Test new url spec. [bee_keeper]

- Merge pull request #32 from mjnaderi/patch-1. [bee-keeper]

  Remove RemovedInDjango110Warning warnings in Django 1.9

- Remove RemovedInDjango110Warning warnings in Django 1.9. [Mohammad
  Javad Naderi]

  With Django 1.9, each time I run `python manage.py runserver`, it shows 2 warnings:

  ```bash
  /someplace/.venv/lib/python2.7/site-packages/invitations/urls.py:15: RemovedInDjango110Warning: django.conf.urls.patterns() is deprecated and will be removed in Django 1.10. Update your urlpatterns to be a list of django.conf.urls.url() instances instead.
    name='accept-invite'),

  /someplace/.venv/lib/python2.7/site-packages/invitations/urls.py:15: RemovedInDjango110Warning: django.conf.urls.patterns() is deprecated and will be removed in Django 1.10. Update your urlpatterns to be a list of django.conf.urls.url() instances instead.
    name='accept-invite'),
  ```

  This pull request fixes it.

## 1.6 (2016-01-10)

- V1.6. [bee_keeper]

- Dont override inviter from create step. [bee_keeper]

- Merge pull request #30 from percipient/allauth-with-custom-adapter.
  [bee-keeper]

  Allow using custom allauth backends.

- Make flake8 happy. [Patrick Cloke]

- Don't return None from get_invitations_adapter. [Patrick Cloke]

- Merge pull request #31 from percipient/set-inviter. [bee-keeper]

  Set the inviter in the SendInvite view.

- Set the inviter in the SendInvite view. [Patrick Cloke]

- Update README.md. [bee-keeper]

## 1.5 (2015-12-07)

- Update README.md. [bee-keeper]

- Merge pull request #24 from bee-keeper/devel. [bee-keeper]

  Refactor as generic invite app

- Removed the dependancy on django-allauth, this package is now a
  generic invite app. [bee_keeper]

## 1.4 (2015-11-27)

- 1.4. [bee_keeper]

- Merge pull request #23 from bee-keeper/devel. [bee-keeper]

  Coverage and exposing inviter in admin

- Coverage and exposing inviter in admin. [bee_keeper]

## 1.3 (2015-11-26)

- Merge pull request #22 from bee-keeper/devel. [bee-keeper]

  Added inviter to invitation

- Added inviter to invitation. [bee_keeper]

- Merge pull request #21 from bee-keeper/devel. [bee-keeper]

  Support for django1.9

- Testing for django1.9 and python 3.5. [bee_keeper]

- Merge pull request #20 from bee-keeper/devel. [bee-keeper]

  Added json endpoint for invites

- Added json endpoint for invites. [bee_keeper]

- Merge pull request #19 from bee-keeper/devel. [bee-keeper]

  Made accept trailing slash optional

- Made trailing slash optional and added flake8 to CI. [bee_keeper]

  Bumped to version 1.3

- Update models.py. [bee-keeper]

- Roadmap. [bee_keeper]

## 1.2 (2015-08-29)

- Test coverage done, ready for 1.2 release. [bee_keeper]

- Dropping support for python 3.2. [bee_keeper]

- Dropping support for python 3.2. [bee_keeper]

- Signal test coverage, tweaking tox. [bee_keeper]

- Coverage. [bee-keeper]

- Tox+travis. [bee-keeper]

- Tox. [bee-keeper]

- Tox+travis. [bee-keeper]

- Testing tox+travis. [bee-keeper]

- Testing tox+travis. [bee-keeper]

- Tox file. [bee_keeper]

- Py3 fix. [bee_keeper]

- Test for signup redirect. [bee_keeper]

- Update README.md. [bee-keeper]

- Py 3.2. [bee_keeper]

- Py 3.2. [bee-keeper]

- Print. [bee-keeper]

- Tests and bug fixes. [bee-keeper]

## 1.1 (2015-08-05)

- V 1.1. [bee_keeper]

- Readme. [bee_keeper]

- Modified PR (15) + editorconfig. [bee_keeper]

- Merge branch 'nwaxiomatic-master' [bee_keeper]

- Admin invitations. [Nic]

  sends invitations from admin on save

## 1.0 (2015-07-26)

- Release 1.0. [bee_keeper]

- Requirements. [bee_keeper]

- Changing travis supported versions. [bee_keeper]

- Travis. [bee_keeper]

- Travis. [bee_keeper]

- Remove 2.6 from testing. [bee_keeper]

- Requirements and changelog. [bee_keeper]

- Test settings. [bee_keeper]

- Requirements.txt. [bee_keeper]

- Travis. [bee_keeper]

- Removing uneeded imports. [bee_keeper]

- Removed ALLOWED_GROUPS setting. [bee_keeper]

- Merge pull request #12 from tbarbugli/patch-1. [bee-keeper]

  fix invite form

- Fix invite form. [Tommaso Barbugli]

- Update views.py. [bee-keeper]

- Teavis. [bee_keeper]

- Travis. [bee_keeper]

- Travis. [bee_keeper]

- Travis. [bee_keeper]

- Travis. [bee_keeper]

- Travis. [bee_keeper]

- App settings. [bee_keeper]

- Merge pull request #6 from simonv3/master. [bee-keeper]

## Redo pull request of adding inviter to signal

- Add reference to inviter in signal. [Simon]

- .travis.yml. [bee_keeper]

- .travis.yml. [bee_keeper]

- Readme. [bee_keeper]


invitations/app_settings.py
from django.conf import settings


class AppSettings(object):
    def __init__(self, prefix):
        self.prefix = prefix

    def _setting(self, name, dflt):
        from django.conf import settings

        return getattr(settings, self.prefix + name, dflt)

    @property
    def INVITATION_EXPIRY(self):
        """How long before the invitation expires"""
        return self._setting("INVITATION_EXPIRY", 3)

    @property
    def INVITATION_ONLY(self):
        """Signup is invite only"""
        return self._setting("INVITATION_ONLY", False)

    @property
    def CONFIRM_INVITE_ON_GET(self):
        """Simple get request confirms invite"""
        return self._setting("CONFIRM_INVITE_ON_GET", True)

    @property
    def ACCEPT_INVITE_AFTER_SIGNUP(self):
        """Accept the invitation after the user finished signup."""
        return self._setting("ACCEPT_INVITE_AFTER_SIGNUP", False)

    @property
    def GONE_ON_ACCEPT_ERROR(self):
        """
        If an invalid/expired/previously accepted key is provided, return a
        HTTP 410 GONE response.
        """
        return self._setting("GONE_ON_ACCEPT_ERROR", True)

    @property
    def ALLOW_JSON_INVITES(self):
        """Exposes json endpoint for mass invite creation"""
        return self._setting("ALLOW_JSON_INVITES", False)

    @property
    def SIGNUP_REDIRECT(self):
        """Where to redirect on email confirm of invite"""
        return self._setting("SIGNUP_REDIRECT", "account_signup")

    @property
    def LOGIN_REDIRECT(self):
        """Where to redirect on an expired or already accepted invite"""
        return self._setting("LOGIN_REDIRECT", settings.LOGIN_URL)

    @property
    def ADAPTER(self):
        """The adapter, setting ACCOUNT_ADAPTER overrides this default"""
        return self._setting("ADAPTER", "invitations.adapters.BaseInvitationsAdapter")

    @property
    def EMAIL_MAX_LENGTH(self):
        """
        Adjust max_length of e-mail addresses
        """
        return self._setting("EMAIL_MAX_LENGTH", 254)

    @property
    def EMAIL_SUBJECT_PREFIX(self):
        """
        Subject-line prefix to use for email messages sent
        """
        return self._setting("EMAIL_SUBJECT_PREFIX", None)

    @property
    def INVITATION_MODEL(self):
        """
        Subject-line prefix to use for Invitation model setup
        """
        return self._setting("INVITATION_MODEL", "invitations.Invitation")

    @property
    def INVITE_FORM(self):
        return self._setting("INVITE_FORM", "invitations.forms.InviteForm")

    @property
    def ADMIN_ADD_FORM(self):
        return self._setting(
            "ADMIN_ADD_FORM", "invitations.forms.InvitationAdminAddForm"
        )

    @property
    def ADMIN_CHANGE_FORM(self):
        return self._setting(
            "ADMIN_CHANGE_FORM", "invitations.forms.InvitationAdminChangeForm"
        )


app_settings = AppSettings("INVITATIONS_")
